### PR TITLE
Feature/115

### PR DESCRIPTION
**Pull request by Redaloukil**
I would set it in WIP for now, because i assume it's not done and need some more improvements, please take a look and tell me if i'm implementing it correctly or not.

Remark i have : 
- The jsonnetfile example you published on the issue doesn't contains enough information as specified in `DependencyManagementFile` interface.

### CODE REVIEW

File: extractor/src/extractors/ExtractorRegistry.ts [hunk 1]
```diff
@@ -11,6 +11,7 @@ import IvyXmlExtractor from "./IvyXmlExtractor";
 import PackageJsonExtractor from "./PackageJsonExtractor";
 import PomXmlExtractor from "./PomXmlExtractor";
 import VendorConfExtractor from "./VendorConfExtractor";
+import JsonnetJsonExtractor from "./JsonnetJsonExtractor";
```
Comment (mjpitz) [inline extractor/src/extractors/ExtractorRegistry.ts:14]: Let's call this `JsonnetfileJsonExtractor` for consistency. I know it's a bit wordy

File: extractor/src/extractors/JsonnetJsonExtractor.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        module:"",
+        versionConstraint: dependency.version,
+        scopes:[scopes],
+        name: parseName(dependency),
+    };
+}
+
+//The name should be formatted as {{ remote }}[#{{ subdir }}] where subdir
+const parseName = (dependency) => {
+    return `${dependency.source.git.remote}[#${dependency.source.git.subdir}]`;
```
Comment (mjpitz) [inline extractor/src/extractors/JsonnetJsonExtractor.ts:19]: Let's remove the `[]`. Most URL libraries will support parsing with a fragment.

File: extractor/src/extractors/JsonnetJsonExtractor.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+            organization,
+            dependencies,
+
+        } = files["jsonnetfile.json"].json();
+
+        const deps = dependencies.map(dependence => extract(dependence , "dependencies"));
+
+
+        return {
+            language: Languages.JAVASCRIPT,
```
Comment (mjpitz) [inline extractor/src/extractors/JsonnetJsonExtractor.ts:49]: Let's create a new 'jsonnet' lang.

File: extractor/src/extractors/JsonnetJsonExtractor.ts [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+            organization,
+            module: "",
+            version,
+            dependencies: deps,
+            name,
+        }
+    }
+}
+
+export default  JsonnetJsonExtractor;
```
Comment (mjpitz) [inline extractor/src/extractors/JsonnetJsonExtractor.ts:61]: Preferred all on one line like the other files

```
export default class Jsonnet..
```

File: extractor/src/extractors/JsonnetJsonExtractor.ts [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    public async extract(url: string, files: { [key: string]: ExtractorFile }): Promise<DependencyManagementFile> {
+        const {
+            name,
+            version,
+            organization,
+            dependencies,
+
+        } = files["jsonnetfile.json"].json();
+
+        const deps = dependencies.map(dependence => extract(dependence , "dependencies"));
```
Comment (mjpitz) [inline extractor/src/extractors/JsonnetJsonExtractor.ts:45]: `dependence` => `dependency` and instead of the `"dependencies"` scope, let's use `"direct"`

File: extractor/src/extractors/JsonnetJsonExtractor.ts [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    }
+    requires(): string[] {
+        return [ "jsonnetfile.json" ];
+    }
+    public async extract(url: string, files: { [key: string]: ExtractorFile }): Promise<DependencyManagementFile> {
+        const {
+            name,
+            version,
+            organization,
+            dependencies,
```
Comment (mjpitz) [inline extractor/src/extractors/JsonnetJsonExtractor.ts:41]: Remove `name`, `version`, and `organization`. `version` is the file format according to the spec.

File: extractor/src/extractors/JsonnetJsonExtractor.ts [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+        return {
+            language: Languages.JAVASCRIPT,
+            system:"",
+            sourceUrl: "",
+            organization,
+            module: "",
+            version,
+            dependencies: deps,
+            name,
```
Comment (mjpitz) [inline extractor/src/extractors/JsonnetJsonExtractor.ts:56]: `name` should be `url`

organization / module / sourceURL shouldn't be set.

File: extractor/src/extractors/JsonnetJsonExtractor.ts [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+            dependencies,
+
+        } = files["jsonnetfile.json"].json();
+
+        const deps = dependencies.map(dependence => extract(dependence , "dependencies"));
+
+
+        return {
+            language: Languages.JAVASCRIPT,
+            system:"",
```
Comment (mjpitz) [inline extractor/src/extractors/JsonnetJsonExtractor.ts:50]: This can be set to `jsonnet-bundler`.

**Review by mjpitz — CHANGES_REQUESTED** [review]

**Comment by mjpitz** [comment]
| The jsonnetfile example you published on the issue doesn't contains enough information as specified in DependencyManagementFile interface.

Correct. We're actually working on simplifying this part of the schema so some of those fields are deprecated. Unfortunately, my NodeJS code generation doesn't support deprecated annotations yet. Still trying to figure out what to do there.

File: extractor/src/extractors/JsonnetJsonExtractor.ts (hunk 7 above)
Comment (Redaloukil) [reply extractor/src/extractors/JsonnetJsonExtractor.ts:56]: `name should be url` what do you mean by url ?

Comment (mjpitz) [reply extractor/src/extractors/JsonnetJsonExtractor.ts:56]: So jsonnet-bundler references libraries using their vcs url and an optional subdir. The way I interpret this into our standardized model is that the module's `name` is the URL of the repo and the `scope` is the `subdir`

File: extractor/src/extractors/JsonnetJsonExtractor.ts (hunk 2 above)
Comment (mjpitz) [reply extractor/src/extractors/JsonnetJsonExtractor.ts:19]: Also, let's make the subdir the scope for the dependency instead of a fragment on the name.

Comment (Redaloukil) [reply extractor/src/extractors/JsonnetJsonExtractor.ts:19]: You mean , I should remove subdir from name  and place in the scopes array of the dependency ?

File: extractor/src/extractors/JsonnetfileJsonExtractor.ts [hunk 9]
```diff
@@ -0,0 +1,40 @@
+import MatchConfig from "../matcher/MatchConfig";
+import Extractor from "./Extractor";
+import ExtractorFile from "./ExtractorFile";
+import Languages from "./Languages";
+
+const extract = (dependency: any,scopes:any) => {
```
Comment (mjpitz) [inline extractor/src/extractors/JsonnetfileJsonExtractor.ts:6]: Remove scopes as a parameter. Just use the subdir if it's set. Otherwise, it should be an empty array.

File: extractor/src/extractors/JsonnetfileJsonExtractor.ts [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+        } = files["jsonnetfile.json"].json();
+
+        const deps = dependencies.map(dependency => extract(dependency , "direct"));
+
+
+        return {
+            language: Languages.JSONNET,
+            system:"jsonnet-bundler",
+            dependencies: deps,
+            name,
```
Comment (mjpitz) [inline extractor/src/extractors/JsonnetfileJsonExtractor.ts:37]: name is not set. should be `name: url,`

**Review by mjpitz — CHANGES_REQUESTED** [review]

File: extractor/src/extractors/JsonnetfileJsonExtractor.ts (hunk 9 above)
Comment (Redaloukil) [reply extractor/src/extractors/JsonnetfileJsonExtractor.ts:6]: Got it , I have a small question about : 
`dependency.source.git.remote` and `dependency.source.git.subdir` should i check all the navigation like : `if ( dependency.source && dependency.source.git && dependency.source.git.remote )` or are we sure`dependency.source.git` is truthy ?

File: extractor/src/extractors/JsonnetfileJsonExtractor.ts [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+import ExtractorFile from "./ExtractorFile";
+import Languages from "./Languages";
+
+const extract = (dependency: any) => {
+    const { subdir , name } = dependency.source.git
+    return {
+        subdir,
+        name,
+    };
+}
```
Comment (mjpitz) [inline extractor/src/extractors/JsonnetfileJsonExtractor.ts:12]: This should return the standard Dependency structure like the others. (see the type annotations)

```
const extract = (dependency: any): Dependency => {
  const { source, version } = dependency;
  const { git } = (source || {});
  const { subdir, name } = (git || {});

  return {
    name,
    versionConstraint: version,
    scopes: subdir ? [ subdir ] : [],
    organization: "", // deprecated
    module: "", // deprecated
  };
};
```

File: extractor/src/extractors/PackageJsonExtractor.ts [hunk 12]
```diff
@@ -20,7 +20,7 @@ function parseName(module: string): ID {
     return { organization, module };
 }
 
-function extract(dependencyHash: any, scope: string): Dependency[] {
+function extract(dependencyHash: any, scope: string): Dependency[] {
```
Comment (mjpitz) [inline extractor/src/extractors/PackageJsonExtractor.ts:23]: unnecessary whitespace change, please revert

File: extractor/src/main.ts [hunk 13]
```diff
@@ -48,11 +48,13 @@ program.name("extractor")
                 agg[item] = true;
                 return agg;
             }, {});
-
+
```
Comment (mjpitz) [inline extractor/src/main.ts:51]: unnecessary

File: extractor/src/main.ts [hunk 14]
```diff
@@ -48,11 +48,13 @@ program.name("extractor")
...
                 return agg;
             }, {});
-
+        
         const extractorReqs = ExtractorRegistry.known()
             .filter((e) => !disabledManifests[e])
             .map((extractor) => ExtractorRegistry.resolve(extractor, null));
 
+        
+
```
Comment (mjpitz) [inline extractor/src/main.ts:57]: unnecessary

File: extractor/src/main.ts [hunk 15]
```diff
@@ -117,6 +119,7 @@ program.name("extractor")
                 timestamp: new Date(),
                 results: {},
             });
+
```
Comment (mjpitz) [inline extractor/src/main.ts:122]: unecessary

File: extractor/src/extractors/JsonnetfileJsonExtractor.ts [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+    }
+    public async extract(url: string, files: { [key: string]: ExtractorFile }): Promise<any> {
+        const {
+           dependencies,
+        } = files["jsonnetfile.json"].json();
+
+        const deps = dependencies.map(dependency => extract(dependency));
+
+
+        return {
```
Comment (mjpitz) [inline extractor/src/extractors/JsonnetfileJsonExtractor.ts:34]: add `organization: "", module: "", // deprecated`

**Comment by mjpitz** [comment]
Thanks @Redaloukil for the PR!
